RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,769 @@
+// be, etc.) and the information needed to generate its help text
+opt {
+  let! username = Opt("username", (Some ""), Some)
+  and! fullname = Opt("fullname", None, Some)
+  and! id = Opt("id", None, readInt)
+  return User(username, fullname, id)
+}
+```
+
+With all of these examples, we can nest the applicative computation expressions inside other computation expressions to build larger and larger descriptions that cleanly separate the pure computation from its context.
+
+# Detailed Design
+[design]: #detailed-design
+
+This RFC introduces a desugaring for applicative computation expressions, much like that which exists for monads and related constructs.
+
+Example desugaring:
+
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    and! z = baz
+    return x + y + z
+ }
+```
+
+becomes
+
+```fsharp
+ce.Apply(
+    ce.Apply(
+        ce.Apply(
+            ce.Return(
+                (fun x ->
+                    (fun y ->
+                        (fun z ->
+                            x + y + z
+                        )
+                    )
+                )),
+            foo),
+        bar),
+    baz)
+```
+
+## Rules on keywords in applicative computation expressions
+
+To be accepted as an applicative computation expression (CE), the CE must be of the form `let! ... and! ... return ...`:
+
+* There must be at least one `and!`s after the `let!`, but there is no hard upper-limit on the number of `and!`s
+* No `let!`s may appear after an `and!` in the same applicative CE
+* No normal `let`s in between the `let!` and `and!`s
+* No usage of `yield` in place of `return`
+* `do!`, `match!` and other CE keywords are also not valid in an applicative CE
+* `use!` and `anduse!` may replace `let!` and `and!`, respectively, to indicate a resource that needs to be managed by the CE builder
+
+### Valid syntax
+
+Only one `and!`:
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar ✔️
+    return x + y
+ }
+```
+
+Many `and!`s:
+```fsharp
+ce {
+    let! w = foo
+    and! x = bar ✔️
+    and! y = baz ✔️
+    and! z = qux ✔️
+    return w + x + y + z
+ }
+```
+
+`let`-binding inside the `return`:
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    and! z = baz️
+    return (let w = x + y in w + z) ✔️
+ }
+```
+
+Function call inside the `return`:
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    and! z = baz
+    return sprintf "x = %d, y = %d, z = %d" x y z ✔️
+ }
+```
+
+The `let!` is replaced by `use!`:
+```fsharp
+ce {
+    use! x = foo ✔️
+    and! y = bar
+    return x + y
+ }
+```
+
+An arbitrary `and!` is replaced by `anduse!`:
+```fsharp
+ce {
+    let!    w = foo
+    and!    x = bar
+    anduse! y = baz ✔️
+    and!    z = qux
+    return w + x + y + z
+ }
+```
+
+The `let! ... and! ...` form is replaced entirely by its resource-tracking equivalent:
+```fsharp
+ce {
+    use!    x = foo ✔️
+    anduse! y = bar ✔️
+    anduse! z = qux ✔️
+    return x + y + z
+ }
+```
+
```

Comment:
Is there  a valid use with combined with `return!`? If so, let's add an example.

Reply:
Thinking about this, I the answer is probably no.